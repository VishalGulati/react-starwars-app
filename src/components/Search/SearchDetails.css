.search-details {
  text-transform: capitalize;
  font-size: 12px;
}