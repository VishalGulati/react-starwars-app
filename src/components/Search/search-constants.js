export const SEARCH_ACTION = 'SEARCH';
export const SEARCH_SUCCESS_ACTION = 'SEARCH_SUCCESS';
export const SEARCH_FAIL_ACTION = 'SEARCH_FAIL';
