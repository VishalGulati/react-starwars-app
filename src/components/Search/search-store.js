import { put, takeLatest, call } from 'redux-saga/effects';
import {
  SEARCH_ACTION,
  SEARCH_SUCCESS_ACTION,
  SEARCH_FAIL_ACTION
} from './search-constants';
import { getSearchResults } from '../../api';

const initialState = {
  isLoading: false,
  searchResults: null,
  error: null
};

export const searchAction = searchTerm => ({
  type: SEARCH_ACTION,
  searchTerm
});

export const searchSuccess = data => ({
  type: SEARCH_SUCCESS_ACTION,
  data
});

export const searchFail = error => ({
  type: SEARCH_FAIL_ACTION,
  error
});

export function* searchSaga(action) {
  try {
    const { searchTerm } = action;
    const data = yield call(getSearchResults, searchTerm);
    yield put(searchSuccess(data.results));
  } catch ({ code, message }) {
    yield put(searchFail({ code, message }));
  }
}

export function* watchSearch() {
  yield takeLatest(SEARCH_ACTION, searchSaga);
}

export default function searchReducer(state = initialState, action) {
  switch (action.type) {
    case SEARCH_ACTION:
      return {
        ...state,
        isLoading: true
      };

    case SEARCH_SUCCESS_ACTION:
      return {
        ...state,
        isLoading: false,
        searchResults: action.data
      };

    case SEARCH_FAIL_ACTION:
      return {
        ...state,
        isLoading: false,
        error: action.error
      };

    default:
      return state;
  }
}
