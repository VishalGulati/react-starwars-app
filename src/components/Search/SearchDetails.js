import React, { Component } from 'react';
import {
    Table
} from 'react-bootstrap';
import './SearchDetails.css';

export default class SearchDetails extends Component {
    render() {
        let searchItemDetails = [];
        if (this.props.location.params) {
            const params = this.props.location.params;
            searchItemDetails = Object.keys(
                params
            ).map((propName, key) => {
                return <tr key={key}>
                    <td>{propName}</td>
                    <td>{params[propName]}</td>
                </tr>
            })
        }

        return (
            <div >
                <h1>Search Details</h1>
                <Table striped bordered condensed hover className="search-details">
                    <tbody>
                        {searchItemDetails}
                    </tbody>
                </Table>
            </div>
        );
    }
}
