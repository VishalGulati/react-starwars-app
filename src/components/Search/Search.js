import React, { Component } from 'react';
import {
  PageHeader,
  Form,
  FormGroup,
  Col,
  FormControl,
  InputGroup,
  Glyphicon,
  ListGroup,
  ListGroupItem
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { searchAction } from './search-store';
import { searchDetailsRoute } from '../../router/route-constants';

export class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: '',
      searchResult: []
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.search !== this.props.search) {
      const { isLoading, error } = nextProps.search;
      if (!isLoading && !error) {

      }
    }
  }

  handleSearch = async e => {
    this.props.searchAction(e.target.value)
  };

  showSearchDetails = (item) => {
    this.props.history.push({
      pathname: `/${searchDetailsRoute}`,
      params: item
    })
  }

  render() {
    let searchResultList = [];
    if (this.props.searchResults) {
      searchResultList = this.props.searchResults.map((item, i) => {
        return <ListGroupItem onClick={() => this.showSearchDetails(item)} key={i}>
          <span>{item.name}</span>
          <span className="pull-right"><Glyphicon glyph="user" /> {item.population}</span>
        </ListGroupItem>
      })
    }

    return (
      <div className="search">
        <PageHeader className="text-center">Search</PageHeader>
        <Form>
          <FormGroup controlId="formUsername">
            <Col sm={12}>
              <InputGroup>
                <FormControl
                  type="text"
                  placeholder="Enter search here ..."
                  autoComplete="off"
                  onChange={this.handleSearch}
                />
                <InputGroup.Addon>
                  <Glyphicon glyph="search" />
                </InputGroup.Addon>
              </InputGroup>
            </Col>
          </FormGroup>
        </Form>

        {searchResultList.length > 0 &&
          <Col sm={12}>
            <ListGroup>
              {searchResultList}
            </ListGroup>
          </Col>
        }
      </div>
    );
  }
}

const mapStateToProps = ({ search: { searchResults } }) => ({
  searchResults
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      searchAction
    },
    dispatch
  );

export default connect(mapStateToProps, mapDispatchToProps)(Search);
