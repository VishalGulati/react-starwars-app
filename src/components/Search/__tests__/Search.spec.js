import React from 'react';
import renderer from 'react-test-renderer';
import { Search } from '../Search';

describe('Search', () => {
    it('should render Search properly', () => {
        const component = renderer.create(<Search />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
