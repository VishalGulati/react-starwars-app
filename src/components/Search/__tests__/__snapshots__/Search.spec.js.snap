// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render Search properly 1`] = `
<div
  className="search"
>
  <div
    className="text-center page-header"
  >
    <h1>
      Search
    </h1>
  </div>
  <form
    className=""
  >
    <div
      className="form-group"
    >
      <div
        className="col-sm-12"
      >
        <span
          className="input-group"
        >
          <input
            className="form-control"
            id="formUsername"
            onChange={[Function]}
            placeholder="Enter search here ..."
            type="text"
          />
          <span
            className="input-group-addon"
          >
            <span
              className="glyphicon glyphicon-search"
            />
          </span>
        </span>
      </div>
    </div>
  </form>
</div>
`;
